### PR TITLE
Service worker: offline page

### PR DESCRIPTION
**Pull request by OliverJAsh**
When you visit a HTTPS page (i.e. https://www.theguardian.com/info or https://www.theguardian.com/info/developer-blog), you will download our service worker file if your browser supports service workers.

The next time you visit one of our HTTPS pages and you're not connected to the internet, you will get an offline page. (See screenshot below, in which I have simulated offline by disabling nginx.)

TODO:
- [ ] Check if I need to make any changes to the nginx router
- [x] Check the cache time for the service worker file (change from no cache)

Async TODOs:
- [x] Explore add to home screen

I may have gone seriously wrong with the Scala. Still learning. Please advise, @rich-nguyen and @janua! :smile: 

image

/cc @jakearchibald @theefer @phamann

### CODE REVIEW

File: common/app/assets/assets.scala [hunk 1]
```diff
@@ -29,27 +29,40 @@ class AssetMap(base: String, assetMap: String) {
...
+    case JsSuccess(myMap, _) => myMap
+      .mapValues(_.mkString)
+      .map(_.swap)
+      .map{case (k, v) => (s"javascripts/$k.js", s"$v.js")}
+    case JsError(errors) => Map.empty
+  }
+
+  val lookup = new AssetMap(base, paths)
+  def apply(path: String): Asset = lookup(path)
+}
```
Comment (OliverJAsh) [inline common/app/assets/assets.scala]: The jspm asset map is in format `Map[BundleFilePath, List[BundledModuleFilePath]]` when we need `Map[BundledModuleFilePath, BundleFilePath]`, hence the munging.

File: applications/app/views/offlinePage.scala.html [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+@import play.api.libs.json._
+
+@main(crosswordPage) { } {
+    <div class="l-side-margins">
+        <article id="crossword" class="content content--article tonal tonal--tone-news" role="main">
+            <div class="content__head tonal__head tonal__head--tone-news">
+                <div class="gs-container">
+                    <div class="content__main-column">
+                        <h1 class="content__headline">Unable to connect to the Internet</h1>
+                        <p>We can't display this webpage because your computer isn't connected to the Internet.</p>
```
Comment (theefer) [inline applications/app/views/offlinePage.scala.html]: Computer? What if I'm on a phone?

**Comment by theefer** [comment]
This is amazing! :christmas_tree:

File: common/app/assets/assets.scala (hunk 1 above)
Comment (OliverJAsh) [inline common/app/assets/assets.scala]: I'm going to make jspm bundle produce the correct format and then reformat for jspm to consume.

File: common/app/templates/systemJsSetup.scala.js [hunk 3]
```diff
@@ -23,8 +23,8 @@ System.config({
 
 @JavaScript(Static.js.systemJsAppConfig)
 
-@if(!play.Play.isDev()) {
-    @JavaScript(Static.js.systemJsBundleConfig)
+@if(!play.Play.isDev() || (play.Play.isDev() && useBundlesInDev)) {
+    System.config({ bundles: @JavaScript(Static.js.systemJsBundleConfig) });
```
Comment (OliverJAsh) [inline common/app/templates/systemJsSetup.scala.js]: Load explicitly in dev on offline page.

**Comment by paperboyo** [comment]
:gem:!

**Comment by slightlyoff** [comment]
SO GOOD

**Comment by jakearchibald** [comment]
Really excited about this. SW looks spot on.

**Comment by jakearchibald** [comment]
In the current version of Chrome, `fetch(request)` will follow redirects and return the eventual non-redirect response. This what you want for the vast majority of requests, but it could cause issues for navigations:

* `/info/developer-blog/2015/aug/27/how-i-went-from-a-tester-to-a-developer-role/` which redirects to
* `/info/developer-blog/2015/aug/27/how-i-went-from-a-tester-to-a-developer-role` (no trailing slash)

Using this PR's SW and the current Chrome, a request to `…how-i-went-from-a-tester-to-a-developer-role/` will swallow the redirect and the url in the url bar will remain `…tester-to-a-developer-role/`. This may impact relative urls, or completely break in cases where the redirect is cross origin (a cross-origin response is not acceptable for a navigation).

This won't be a problem once https://code.google.com/p/chromium/issues/detail?id=522873 is fixed. Navigation requests will have a redirect mode of "manual", which means `fetch()` won't swallow the redirect and everything will behave as expected.

In the mean time, you can either:

## Not care

If you don't have any redirects that will be broken by this, all is well, you can ignore it.

## Check the response url and act accordingly 

```js
this.addEventListener('fetch', function (event) {
    // this assumes we invalidate the cache every release!
    event.respondWith(
        fetch(event.request)
            .then(function(response) {
                if (response.url != request.url) {
                    return Response.redirect(response.url);
                }
                return response;
            })
            .catch(function () {
                // If a request is cached, respond with that. Otherwise respond with the shell, whose subresources will
                // in the cache.
                return caches.match(event.request).then(function (response) {
                    return response || caches.match('/offline-page');
                })
            })
    );
});
```

Unfortunately this means redirects end up with an extra response. One for the `fetch()`, and another when the navigation handles the redirect. For same-origin redirects you can hack around this:

```js
var tmpCache = {};

this.addEventListener('fetch', function (event) {
    // this assumes we invalidate the cache every release!
    if (tmpCache[event.request.url]) {
        var response = tmpCache[event.request.url];
        delete tmpCache[event.request.url];
        event.respondWith(response);
        return;
    }

    event.respondWith(
        fetch(event.request)
            .then(function(response) {
                if (response.url != request.url) {
                    tmpCache[response.url] = response;
                    return Response.redirect(response.url);
                }
                return response;
            })
            .catch(function () {
                // If a request is cached, respond with that. Otherwise respond with the shell, whose subresources will
                // in the cache.
                return caches.match(event.request).then(function (response) {
                    return response || caches.match('/offline-page');
                })
            })
    );
});
```

## Wait until Chrome 46 and use manual redirects

```js
this.addEventListener('fetch', function (event) {
    event.respondWith(
        fetch(event.request, {
            redirect: 'manual'
        }).catch(function () {
            // If a request is cached, respond with that. Otherwise respond with the shell, whose subresources will
            // in the cache.
            return caches.match(event.request).then(function (response) {
                return response || caches.match('/offline-page');
            })
        })
    );
});
```

This will use a manual redirect for everything, meaning `fetch()` will return redirect responses.

## Leave code as it is, and wait for https://code.google.com/p/chromium/issues/detail?id=522873 to be fixed (likely Chrome 47)

After this change, navigation requests will automatically have a "manual" redirect mode.

File: common/app/assets/assets.scala (hunk 1 above)
Comment (OliverJAsh) [inline common/app/assets/assets.scala]: @rich-nguyen See 496e16a2e3ef3bd3d71aacd440409a671716ca79

File: common/app/templates/systemJsSetup.scala.js (hunk 3 above)
Comment (OliverJAsh) [inline common/app/templates/systemJsSetup.scala.js]: I can't make this change because the head you pass along to main is inserted before System itself.

**Comment by OliverJAsh** [comment]
Thanks for the low-down on that, @jakearchibald.

I *think* we only ever link to `/path`, never `/path/`. We might be linking to the latter in some places, but if they are few and far between then I'm happy to leave those broken until navigation requests automatically have a "manual" redirect mode.

File: applications/app/services/ContentApiGetters.scala [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+object ContentApiGetters {
+  def withCrossword(crosswordType: String, id: Int)(f: (Crossword, ApiContent) => Result)(implicit request: RequestHeader): Future[Result] = {
+    LiveContentApi.getResponse(LiveContentApi.item(s"crosswords/$crosswordType/$id", Edition(request)).showFields("all")).map { response =>
+      val maybeCrossword = for {
+        content <- response.content
+        crossword <- content.crossword }
+        yield f(crossword, content)
+      maybeCrossword getOrElse InternalServerError("Crossword response from Content API invalid.")
+    } recover { case _ => InternalServerError("Content API query returned an error.") }
```
Comment (janua) [inline applications/app/services/ContentApiGetters.scala]: You might want to log here. If this `Future` starts failing the only way to tell what happened is to actually change the code and deploy, but if you are logging it out you will be able to spot it in the logs.

File: applications/app/controllers/CrosswordsController.scala [hunk 5]
```diff
@@ -1,30 +1,21 @@
...
-import play.api.mvc.{Action, Controller, RequestHeader, Result, _}
+import play.api.mvc.{Action, Controller, _}
 import services.IndexPage
+import services.ContentApiGetters.withCrossword
 
 import scala.concurrent.Future
 import scala.concurrent.duration._
 
 object CrosswordsController extends Controller with ExecutionContexts {
-  protected def withCrossword(crosswordType: String, id: Int)(f: (Crossword, ApiContent) => Result)(implicit request: RequestHeader): Future[Result] = {
```
Comment (janua) [inline applications/app/controllers/CrosswordsController.scala]: Why did you move this to it's own `object`?

File: applications/app/controllers/WebAppController.scala [hunk 6]
```diff
@@ -18,4 +30,15 @@ object WebAppController extends Controller with ExecutionContexts {
   def manifest() = Action {
     Cached(3600) { Ok(templates.js.webAppManifest()) }
   }
+
+  def offlinePage() = Action.async { implicit request =>
+    if (conf.Switches.OfflinePageSwitch.isSwitchedOn) {
+      withCrossword("quick", 14127) { (crossword, content) =>
```
Comment (janua) [inline applications/app/controllers/WebAppController.scala]: I presume you are planning on making the crossword different at some point!?

File: applications/app/services/ContentApiGetters.scala (hunk 4 above)
Comment (OliverJAsh) [inline applications/app/services/ContentApiGetters.scala]: See 964b48492a898091045c7d8a3f09a4e7fc1ccc73

File: applications/app/controllers/WebAppController.scala (hunk 6 above)
Comment (OliverJAsh) [inline applications/app/controllers/WebAppController.scala]: Yup, still figuring out how to make that work with service workers. Not sure yet…

File: applications/app/controllers/CrosswordsController.scala (hunk 5 above)
Comment (OliverJAsh) [inline applications/app/controllers/CrosswordsController.scala]: So it can be shared between `CrosswordsController` and `WebAppController`.

Comment (janua) [inline applications/app/controllers/CrosswordsController.scala]: Makes sense. Although you could have still done it straight off the controller if you needed!

**Comment by OliverJAsh** [comment]
Changes:

* Use the latest quick crossword and vary cache by date. (New crosswords are published at midnight.) https://github.com/guardian/frontend/commit/2ec5d0938e0ee53f7a26ef5b4be5984d7fcd1aca. Spoke to @jakearchibald about ways to update and settled on this, but I'm not 100% confident I've got it right. Jake, do you have any thoughts?
* Hide crossword, add button to show https://github.com/guardian/frontend/commit/bdc5c811f2b0986ab26ce59b318a9a634b56e5c5

image

After clicking on the button:
image

File: common/app/assets/assets.scala [hunk 7]
```diff
@@ -142,17 +142,21 @@ class Assets(base: String, assetMap: String = "assets/assets.map") extends Loggi
   }
 
   object js {
-     private def inlineJs(path: String): String = IOUtils.toString(AssetFinder(path))
+     private def contents(path: String): String = IOUtils.toString(AssetFinder(path))
```
Comment (rich-nguyen) [inline common/app/assets/assets.scala]: why does this rename help? all of those objects are inlines.

File: common/app/assets/assets.scala [hunk 8]
```diff
@@ -142,17 +142,21 @@ class Assets(base: String, assetMap: String = "assets/assets.map") extends Loggi
...
-     val systemJsAppConfig: String = inlineJs("assets/systemjs-config.js")
-     val systemJsNormalize: String = inlineJs("assets/systemjs-normalize.js")
-     val systemJsBundleConfig: String = inlineJs("assets/systemjs-bundle-config.js")
+     val systemJsPolyfills: String = contents("assets/system-polyfills.src.js")
+     val systemJs: String = contents("assets/system.src.js")
+     val systemJsAppConfig: String = contents("assets/systemjs-config.js")
+     val systemJsNormalize: String = contents("assets/systemjs-normalize.js")
 
-     lazy val systemJsSetupFragment: String = templates.js.systemJsSetup().body
+     private val map: Map[String, List[String]] = Json.parse(contents("assets/jspm-assets.map")).validate[Map[String, String]] match {
```
Comment (rich-nguyen) [inline common/app/assets/assets.scala]: rename `map` to something more specific, like `jspmAssetMap`

File: common/app/assets/assets.scala (hunk 7 above)
Comment (OliverJAsh) [inline common/app/assets/assets.scala]: The usage of this function on [line 154](https://github.com/guardian/frontend/pull/10327/files#diff-aef18c02671cecab73288a55f9311a39R154) isn't for inlining JS, it's simply used for reading the file contents (and then parsing that as JSON).

Comment (rich-nguyen) [inline common/app/assets/assets.scala]: Then it should be in a different object.

Comment (rich-nguyen) [inline common/app/assets/assets.scala]: A separate one for inlines, another for jspm config?

File: static/src/javascripts/projects/common/modules/preferences/main.js [hunk 9]
```diff
@@ -207,7 +207,7 @@ define([
                     }
                 });
 
-                navigator.serviceWorker.register('/2015-05-28-2-service-worker.js')
+                navigator.serviceWorker.register('/service-worker.js')
```
Comment (rich-nguyen) [inline static/src/javascripts/projects/common/modules/preferences/main.js]: :+1:

File: common/app/templates/headerInlineJS/bootSystemJS.scala.js [hunk 10]
```diff
@@ -3,7 +3,8 @@
 @import conf.Static
 @import conf.Configuration
 
-@JavaScript(Static.js.systemJsSetupFragment)
+// Service worker only caches the bundles, so the offline page can only ask for those
+@JavaScript(templates.js.systemJsSetup(useBundlesInDev = item.id == "offline-page").body)
```
Comment (rich-nguyen) [inline common/app/templates/headerInlineJS/bootSystemJS.scala.js]: This still seems really odd. Maybe another way to do this is to avoid using `main.scala.html` entirely. For one thing, a number of those fragments really won't work offline anyway. Then you can configure the bundles, etc however you like.

**Comment by rich-nguyen** [comment]
:+1: very awesome. Much simpler templates without using the common fragments, I think.

**Comment by OliverJAsh** [comment]
Going to deploy this on Monday because we're having issues with TeamCity.
